core/services/flux_monitor.go

```diff
@@ -122,7 +122,9 @@ func (fm *concreteFluxMonitor) Disconnect() {
 
 // Disconnect cleans up running deviation checkers.
 func (fm *concreteFluxMonitor) Stop() {
-	fm.cancel()
+	if fm.cancel != nil {
```

Comment:
I personally prefer to make sure the code doesn't call Stop without calling Start, rather than having code try to make up for my own programming mistakes. Would that be practical?

Reply:
Maybe. It would mean modifying all the tests to call `Start()` when they just want to instantiate and then cleanup:

```
app, cleanup := cltest.NewApp()
// need to start
defer cleanup
// breaks
```

seems onerous tho

Reply:
But `NewApp` could just call `Start` right?

Reply:
Pulling in `App` without calling `Start` feels like an anti pattern to me.

Reply:
I agree with this in principle.  Only thing is, I've noticed that there are cases in the tests (due to some technical debt in `cltest`) where it's not always feasible to rely on that.  The cleanup/cancel/close/stop code ends up called twice, which is why I wound up adding some code like this (and also `sync.Once`) in a few places.

Reply:
I don't think it's that straightforward. Especially with all the tests we have using `NewApp` and then configuring mocks before Start.

Reply:
If you want, I can just tweak the tests in question and leave it as is. I'm indifferent.

Reply:
Discussed offline, will recap here.

The issue exposed by the call stack below was that there was an errror during `Start()` that kicked off a cleanup routine resulting in a call to `Stop()` that indiscriminately called Stop to every service. The stance I believe we're all comfortable with is that it's easier to call `Stop()` on everything than cherry pick which started instances to stop during cleanup in an error state.

The added bonus of this due diligence is that we discovered the root cause of the cleanup routine being kicked off: parallel tests on the global `gracefulpanic`. Those `t.Parallel` have since been removed.

Call stack:

```
?   	chainlink/core/store/migrations/migration1573812490	[no test files]
?   	chainlink/core/store/migrations/migration1575036327	[no test files]
?   	chainlink/core/store/migrations/migration1576022702	[no test files]
ok  	chainlink/core/store/models	19.942s
2019-12-19T16:01:13Z [ERROR] unable to lock ORM: postgres advisory locking strategy failed on .Lock, timeout set to 500ms: pq: canceling statement due to user request: can't acquire advisory lock logger/logger.go:190    stacktrace=chainlink/core/logger.Error
	/root/project/core/logger/logger.go:159
chainlink/core/logger.Errorf
	/root/project/core/logger/logger.go:190
chainlink/core/store/orm.(*ORM).MustEnsureAdvisoryLock
	/root/project/core/store/orm/orm.go:111
chainlink/core/store/orm.(*ORM).RawDB
	/root/project/core/store/orm/orm.go:1174
chainlink/core/store/orm_test.TestPostgresLockingStrategy_CanBeReacquiredByNewNodeAfterDisconnect.func2
	/root/project/core/store/orm/locking_strategies_test.go:150
github.com/stretchr/testify/assert.didPanic.func1
	/go/pkg/mod/github.com/stretchr/testify@v1.4.0/assert/assertions.go:917
github.com/stretchr/testify/assert.didPanic
	/go/pkg/mod/github.com/stretchr/testify@v1.4.0/assert/assertions.go:919
github.com/stretchr/testify/assert.Panics
	/go/pkg/mod/github.com/stretchr/testify@v1.4.0/assert/assertions.go:933
github.com/stretchr/testify/require.Panics
	/go/pkg/mod/github.com/stretchr/testify@v1.4.0/require/require.go:1226
chainlink/core/store/orm_test.TestPostgresLockingStrategy_CanBeReacquiredByNewNodeAfterDisconnect
	/root/project/core/store/orm/locking_strategies_test.go:149
testing.tRunner
	/usr/local/go/src/testing/testing.go:865 
--- FAIL: TestPostgresLockingStrategy_CanBeReacquiredByNewNodeAfterDisconnect (1.31s)
    locking_strategies_test.go:149: 
        	Error Trace:	locking_strategies_test.go:149
        	Error:      	func (assert.PanicTestFunc)(0xdec970) should panic
        	            		Panic value:	<nil>
        	Test:       	TestPostgresLockingStrategy_CanBeReacquiredByNewNodeAfterDisconnect
2019-12-19T16:01:17Z [ERROR] unable to lock ORM: postgres advisory locking strategy failed on .Lock, timeout set to 500ms: pq: canceling statement due to user request: can't acquire advisory lock logger/logger.go:190    stacktrace=chainlink/core/logger.Error
	/root/project/core/logger/logger.go:159
chainlink/core/logger.Errorf
	/root/project/core/logger/logger.go:190
chainlink/core/store/orm.(*ORM).MustEnsureAdvisoryLock
	/root/project/core/store/orm/orm.go:111
chainlink/core/store/orm.(*ORM).RawDB
	/root/project/core/store/orm/orm.go:1174
chainlink/core/store/orm_test.TestPostgresLockingStrategy_WhenReacquiredOriginalNodeErrors.func1
	/root/project/core/store/orm/locking_strategies_test.go:178
github.com/stretchr/testify/assert.didPanic.func1
	/go/pkg/mod/github.com/stretchr/testify@v1.4.0/assert/assertions.go:917
github.com/stretchr/testify/assert.didPanic
	/go/pkg/mod/github.com/stretchr/testify@v1.4.0/assert/assertions.go:919
github.com/stretchr/testify/assert.Panics
	/go/pkg/mod/github.com/stretchr/testify@v1.4.0/assert/assertions.go:933
github.com/stretchr/testify/require.Panics
	/go/pkg/mod/github.com/stretchr/testify@v1.4.0/require/require.go:1226
chainlink/core/store/orm_test.TestPostgresLockingStrategy_WhenReacquiredOriginalNodeErrors
	/root/project/core/store/orm/locking_strategies_test.go:177
testing.tRunner
	/usr/local/go/src/testing/testing.go:865 
--- FAIL: TestPostgresLockingStrategy_WhenReacquiredOriginalNodeErrors (1.09s)
    locking_strategies_test.go:177: 
        	Error Trace:	locking_strategies_test.go:177
        	Error:      	func (assert.PanicTestFunc)(0xdec9d0) should panic
        	            		Panic value:	<nil>
        	Test:       	TestPostgresLockingStrategy_WhenReacquiredOriginalNodeErrors
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd29ed4]

goroutine 10600 [running]:
chainlink/core/services.(*concreteFluxMonitor).Stop(0xc00018bf80)
	/root/project/core/services/flux_monitor.go:125 +0x24
chainlink/core/services.(*ChainlinkApplication).Stop.func1()
	/root/project/core/services/application.go:142 +0x154
sync.(*Once).Do(0xc00053b6f8, 0xc00076aee0)
	/usr/local/go/src/sync/once.go:44 +0xb3
chainlink/core/services.(*ChainlinkApplication).Stop(0xc00053b680, 0xc00076af58, 0x2)
	/root/project/core/services/application.go:136 +0x6c
chainlink/core/services.(*ChainlinkApplication).Start.func1(0xc0007d2060, 0xc00053b680)
	/root/project/core/services/application.go:110 +0xc5
created by chainlink/core/services.(*ChainlinkApplication).Start
	/root/project/core/services/application.go:105 +0xf7
FAIL	chainlink/core/store/orm	27.038s
ok  	chainlink/core/store/presenters	0.029s
ok  	chainlink/core/utils	0.034s
ok  	chainlink/core/web	131.993s
?   	chainlink/operator_ui	[no test files]
```